disasm-test/Main.hs

```diff
@@ -318,119 +361,200 @@ cube = TS.mkCUBE
+-- and prints the difference between the parsed version and the .ll file.
+-- Returns the library parsed version and the serialized AST from the library.
 
 processLL :: FilePath -> FilePath -> TestM (FilePath, Maybe FilePath)
 processLL pfx f = do
-  Details dets <- gets showDetails
-  when dets $ liftIO $ putStrLn (showString f ": ")
-  X.handle logError $
-    withFile (assembleToBitCode pfx f)  $ \ bc   ->
-    withFile (disasmBitCode pfx bc) $ \ norm -> do
-      (parsed, ast) <- processBitCode pfx bc
-      when dets $ liftIO $ do
-        ignore (Proc.callProcess "diff" ["-u", norm, parsed])
-        putStrLn ("successfully parsed " ++ show f)
-      return (parsed, ast)
+  Details det <- gets showDetails
+  when det $ liftIO $ putStrLn (showString f ": ")
+  X.handle logError
+    $ withFile (assembleToBitCode pfx f)
+    $ parseBC pfx
   where
     logError (ParseError msg) =
       liftIO $ assertFailure $ unlines
       $ "failure" : map ("; " ++) (lines (formatError msg))
 
+parseBC :: FilePath -> FilePath -> TestM (FilePath, Maybe FilePath)
+parseBC pfx bc = do
+  withFile (disasmBitCode pfx bc) $ \ norm -> do
+    (parsed, ast) <- processBitCode pfx bc
+    Details dets <- gets showDetails
+    when dets $ liftIO $ do
+      -- Informationally display if there are differences between the llvm-dis
+      -- and llvm-disasm outputs, but no error if they differ.
+      ignore (Proc.callProcess "diff" ["-u", norm, parsed])
+      putStrLn ("successfully parsed " ++ show pfx ++ " bitcode")
+    return (parsed, ast)
+
+----------------------------------------------------------------------
+-- Compiler->Assembly->Disassembly tests
+
+-- The compilation tests ensure that the clang version-specific generated .bc
+-- file can be reasonably parsed by this library.  This is a parallel to the
+-- assemblyCube-driven tests, but starts with a C source file.  One distinction
+-- is that the .ll used for the assemblyCube is typically representative of a
+-- specific LLVM version, and while it is assembled and disassembled by newer
+-- versions of LLVM tools, it will never introduce any newer element, whereas the
+-- clang-generated bitcode will contain version-current output which might have
+-- newer elements and ordering.
+--
+-- The compilerCube uses .c files as the input, and .ll files for the expected
+-- output.  The assemblyCube uses the .ll file as both input and output.  The
+-- actual testing done is very similar, and the assemblyCube always generates a
+-- superset of the compilerCube tests (i.e. when no .c file is present).
+
+compilerCube :: TS.CUBE
+compilerCube = assemblyCube
+               { TS.rootName = "*.c"
+               , TS.sweetAdjuster = rangeMatch
+               }
+
+
+runCompileTest :: TS.Sweets -> TS.Expectation -> IO [TestTree]
+runCompileTest sweet expct = do
+  skipTest <- ("SKIP_TEST" `L.isPrefixOf`) <$> L.readFile (TS.expectedFile expct)
```

Comment:
I'm slightly bothered by the fact that this magic string `"SKIP_TEST"` is now used in multiple locations. Can you factor out the `"SKIP_TEST"` logic into its own function?

Reply:
Good idea: done.